flecsi/data/hpx/policy.hh

```diff
@@ -0,0 +1,396 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+#include <hpx/modules/futures.hpp>
+#include <hpx/modules/serialization.hpp>
+
+#include "flecsi/data/field_info.hh"
+#include "flecsi/run/backend.hh"
+#include "flecsi/util/array_ref.hh"
+#include "flecsi/util/hpx.hh"
+
+#include <algorithm>
+#include <cstddef>
+#include <numeric>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace flecsi {
+namespace data {
+
+namespace hpx {
+struct region_impl {
+  // The constructor is collectively called on all ranks with the same s,
+  // and fs. s.first is number of rows while s.second is number of columns.
+  // MPI may assume "number of rows" == number of ranks. The number of columns
+  // is a placeholder for the number of data points to be stored in a
+  // particular row (aka rank), it could be exact or in the case when we don't
+  // know the exact number yet, a large number (flecsi::data::logical_size) is
+  // used.
+  //
+  // Generic frontend code supplies `s` and `fs` (for information about fields),
+  // requesting memory to be (notionally) reserved from backend. Here we only
+  // create the underlying std::vector<> without actually allocating any memory.
+  // The client code (mostly exec::task_prologue) will call .get_storage() with
+  // a field id and the number of elements on this rank, as determined by
+  // partitioning of the field. We will then call .resize() on the
+  // std::vector<>.
+  region_impl(size2 s, const fields & fs) : s(std::move(s)), fs(fs) {
```

Comment:
(I'm writing this in just one place.)

There is a _lot_ of copy-paste from the MPI backend here.  It's fine if HPX can use similar or even the same types, but we definitely shouldn't be duplicating the code (not least because it's difficult to meaningfully review it).  We could consider adding a `data/local` directory to hold files that are used for more than one backend that can allocate memory directly.

Reply:
Yes, the data `submodule` is still heavily WIP and requires more work. So far I did the minimal necessary work to get things going. I will try to unify the two backends, if possible,

Reply:
@opensdh a general question: would you rather like for PRs to be smaller or should I strive for finishing all of the functionalities (and related refactorings) before creating them? I personally would rather have the PRs smaller and more self-contained. For this reason I would like to move resolving this comment (and the one below about code duplication) to a separate PR and focus here on making the HPX backend functional in a minimal way. But I will try to adapt to the group's expectations. Please let me know.

Reply:
First: I'm not in charge of policy here.  That said, my preference is generally for pull requests to be smaller&mdash;but that's largely in terms of the amount of code to review and the number of _unrelated_, complex changes.  On the other hand, commits should be "complete", each should, if at all possible, build and pass tests, and should not change lines that we _expect_ to be changed again in related subsequent commits.  This does sometimes make new features monolithic, since there's not much point in adding a bunch of dead code in one commit and then adding its clients (perhaps unit tests) in another.  When that creates a large review task, that's unfortunate for scheduling reasons, but it's _better_ to review such things all at once rather than try to regain the mental state for each piece.

My concern here is a mixture of things: including large quantities of "MPI code" (scare quotes because we might expect it to become MPI/HPX code after deduplication) here interferes with review and goes against the principle of not introducing changes that we expect to "revert".  The review is made more difficult not only because one would have to check out this version and `diff` it against the MPI version to _find_ the "real changes", but also because one would have to guess what changes were likely in the future in order to interpret the intended architecture and assess whether the duplication were "temporarily necessary".  The duplication is also a form of technical debt&mdash;if we make changes to the MPI backend between merging this and some future deduplication, that deduplication becomes more expensive and/or we have to consider whether to update this copy at the same time.

Reply:
I agree with what you said here and will try to submit PRs accordingly. 

As for the initial issue of duplicating too much code from the data submodule, I think the code (PR) has changed sufficiently with my later commits that may make it more difficult to re-use. HPX requires additional data members that will pervasively change the originally copied structures.

My focus will be on getting functionalities in place first while doing refactorings as needed in a second step.

Reply:
This has been resolved now by factoring out the common code into data/local/policy.hh